libs/ballot-interpreter-nh-next/src/timing_marks.rs

```diff
@@ -0,0 +1,1038 @@
+
+    /// The bounds of the oval mark in the scanned source image that was
+    /// determined to be the best match.
+    pub matched_bounds: Rect,
+
+    /// The cropped source image at `matched_bounds`.
+    #[serde(skip_serializing)]
+    pub source_image: GrayImage,
+
+    /// The cropped source image at `matched_bounds` with each pixel binarized
+    /// to either 0 (black) or 255 (white).
+    #[serde(skip_serializing)]
+    pub binarized_source_image: GrayImage,
+
+    /// A binarized diff image of `binarized_source_image` with the template.
+    /// The more white pixels, the better the match.
+    #[serde(skip_serializing)]
+    pub match_diff_image: GrayImage,
+
+    /// A binarized diff image of `binarized_source_image` with the fill of the
+    /// template. The more black pixels, the better the fill.
+    #[serde(skip_serializing)]
+    pub fill_diff_image: GrayImage,
+}
+
+impl std::fmt::Debug for ScoredOvalMark {
+    fn fmt(&self, f: &mut Formatter) -> std::fmt::Result {
+        write!(
+            f,
+            "ScoredOvalMark {{ location: {:?}, match_score: {}, fill_score: {}, matched_bounds: {:?} }}",
+            self.location, self.match_score, self.fill_score, self.matched_bounds
+        )
+    }
+}
+
+pub const DEFAULT_MAXIMUM_SEARCH_DISTANCE: u32 = 7;
+
+pub type ScoredOvalMarks = Vec<(GridPosition, Option<ScoredOvalMark>)>;
+
+#[time]
+pub fn score_oval_marks_from_grid_layout(
+    img: &GrayImage,
+    oval_template: &GrayImage,
+    timing_mark_grid: &TimingMarkGrid,
+    grid_layout: &GridLayout,
+    side: BallotSide,
+    debug: &ImageDebugWriter,
+) -> ScoredOvalMarks {
+    let threshold = otsu_level(img);
+
+    let scored_ovals = &grid_layout
+        .grid_positions
+        .par_iter()
+        .flat_map(|grid_position| {
+            let location = grid_position.location();
+
+            if location.side != side {
+                return vec![];
+            }
+
+            timing_mark_grid
+                .point_for_location(location.column, location.row)
+                .map_or_else(
+                    || vec![(grid_position.clone(), None)],
+                    |expected_oval_center| {
+                        vec![(
+                            grid_position.clone(),
+                            score_oval_mark(
+                                img,
+                                oval_template,
+                                expected_oval_center,
+                                &location,
+                                DEFAULT_MAXIMUM_SEARCH_DISTANCE,
+                                threshold,
+                            ),
+                        )]
+                    },
+                )
+        })
+        .collect::<ScoredOvalMarks>();
+
+    debug.write("scored_oval_marks", |canvas| {
+        debug::draw_scored_oval_marks_debug_image_mut(canvas, scored_ovals);
+    });
+
+    scored_ovals.clone()
+}
+
+/// Scores an oval mark within a scanned ballot image.
```

Comment:
Could use a bit of commentary explaining the algorithm

Reply:
Done.